bsp/espressif/esp/src/hals/esp32-c3/gpio.zig

```diff
@@ -0,0 +1,148 @@
+            .INV_SEL = @intFromBool(config.output_invert),
+            .OEN_INV_SEL = 0,
+            .padding = 0,
+        });
+
+        GPIO.PIN[number].modify(.{
+            .PIN_PAD_DRIVER = @intFromBool(config.open_drain),
+        });
+
+        // NOTE: Assert that the USB_SERIAL_JTAG peripheral which uses pins GPIO18 and GPIO19
+        //       is disabled and USB pullup/down resistors are disabled.
+        if (number == 18 or number == 19) {
+            const usb_conf0 = peripherals.USB_DEVICE.CONF0.read();
+            std.debug.assert(usb_conf0.USB_PAD_ENABLE == 0 and usb_conf0.DP_PULLUP == 0 and usb_conf0.DP_PULLDOWN == 0 and usb_conf0.DM_PULLUP == 0 and usb_conf0.DM_PULLDOWN == 0);
+        }
+
+        GPIO.ENABLE_W1TS.write(.{ .ENABLE_W1TS = @as(u17, 1) << number, .padding = 0 });
+
+        return Pin{
+            .number = number,
+        };
+    }
+
+    pub fn deinit(self: Pin) void {
+        IO_MUX.GPIO[self.number].write_raw(0x00);
+        GPIO.FUNC_OUT_SEL_CFG[self.number].write_raw(0x00);
+        GPIO.PIN[self.number].write_raw(0x00);
+        GPIO.ENABLE_W1TC.write(.{ .ENABLE_W1TC = @as(u17, 1) << self.number, .padding = 0 });
+    }
+
+    pub fn set_output_enable(self: Pin, enable: bool) void {
+        GPIO.FUNC_OUT_SEL_CFG[self.number].modify(.{
+            .OEN_SEL = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_open_drain(self: Pin, enable: bool) void {
+        GPIO.PIN[self.number].modify(.{
+            .PIN_PAD_DRIVER = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_pullup(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .MCU_WPU = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_pulldown(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .MCU_WPD = @intFromBool(enable),
+        });
+    }
+
+    pub fn set_output_invert(self: Pin, enable: bool) void {
+        GPIO.FUNC_OUT_SEL_CFG[self.number].modify(.{
+            .OEN_INV_SEL = enable,
+        });
+    }
+
+    pub fn set_output_drive_strength(self: Pin, strength: DriveStrength) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .FUN_DRV = @intFromEnum(strength),
+        });
+    }
+
+    pub fn set_input_filter(self: Pin, enable: bool) void {
+        IO_MUX.GPIO[self.number].modify(.{
+            .FILTER_EN = @intFromBool(enable),
+        });
+    }
+
+    pub fn write(self: Pin, level: Level) void {
+        std.debug.assert(GPIO.FUNC_OUT_SEL_CFG[self.number].read().OEN_SEL == 1);
+
+        switch (level) {
+            Level.low => GPIO.OUT_W1TC.write(.{ .OUT_W1TC = @as(u17, 1) << self.number, .padding = 0 }),
+            Level.high => GPIO.OUT_W1TS.write(.{ .OUT_W1TS = @as(u17, 1) << self.number, .padding = 0 }),
+        }
+    }
+
+    pub fn get_output_state(self: Pin) Level {
+        std.debug.assert(GPIO.FUNC_OUT_SEL_CFG[self.number].read().OEN_SEL == 1);
```

Comment:
Similar sentiment to previous comment, a lot of these asserts I think can just be deleted and it's up to the user to correctly configure the pin!